### PR TITLE
Remove hook proxy cache 

### LINKED ISSUE #2016: pytest_ignore_collect applies to all directories collected afterwards?
I put a pytest_ignore_collect() in a conftest.py. From the documentation, I had expected it to apply to all subdirectories of the one in which the conftest.py resides. However, it also prevented collection of files that were not in subdirectories!

I have this directory structure:

demo_parent/childa/conftest.py
demo_parent/childa/demo_test.py
demo_parent/childb/demo_test.py

demo_test.py are both `def test(): pass`. When I add `def pytest_ignore_collect(path, config): return True` to the conftest.py, neither of the tests are executed! I would have expected childb/demo_test.py to run.

platform darwin -- Python 2.7.10, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
plugins: hypothesis-3.4.0, cov-2.2.1, faulthandler-1.3.0, xdist-1.14

also:
execnet==1.4.1
hypothesis==3.4.0
coverage==4.0.3

I've observed this on Linux, and Mac.

### PR DESCRIPTION
**Pull request by nicoddemus**
Fixes #2016 

@hpk42 I would appreciate if you could take a look as you probably know the most about that part of the code. [This comment](https://github.com/pytest-dev/pytest/issues/2016#issuecomment-261940763) contains the full description of my diagnosis of the problem.

cc @d-b-w

### CODE REVIEW

File: _pytest/pytester.py [hunk 1]
```diff
@@ -481,6 +481,7 @@ def _makefile(self, ext, args, kwargs):
         ret = None
         for name, value in items:
             p = self.tmpdir.join(name).new(ext=ext)
+            p.dirpath().ensure_dir()
```
Comment (d-b-w) [inline _pytest/pytester.py]: This doesn't seem wrong or anything, but is it actually relevant to this code change?

Comment (nicoddemus) [inline _pytest/pytester.py]: No, just that I needed this to make the test easier. Now one can use sub-directories in the keys to `testdir.makepyfile` (actually any "make*" functions):

```python
testdir.makepyfile(**{
    'foo/bar/foobar.py': 'print("hello world")',
})
```

**Comment by nicoddemus** [comment]
>  But I do want to double check, given that this is an increase in complexity - did you consider just removing the _fs2hookproxy cache? My (very limited) experiments show that the cache doesn't improve performance at all.

I did consider, and I thought that since `gethookproxy` is called for each entry in a directory it would impact collection considerably.

So I decided to make a benchmark to see if it was true. I wrote [this script](https://gist.github.com/nicoddemus/7ab1d515c6b758ec1b4e466503e7d8a4) which creates an arbitrarily nested directory tree with tests on it. Running:

```
generate_tests_tree.py 4 6 10 0
```

Generates a tree 4 levels deep, 6 folders per level and 10 test files each, without any conftest.py files (that's what the last `0` means). Each test file contains 10 dummy tests, which just `pass`.

Here are the results I got on my Windows box:

```

no-conftests:
-------------

python generate_tests_tree.py 4 10 5 0
2000 tests
cache: 2,0055171
no-cache: 2,0317655


python generate_tests_tree.py 4 5 10 0 
2000 tests
cache: 1,9863132
no-cache: 2,0168685

python generate_tests_tree.py 10 10 30 0
30000 tests
cache: 16,5217091
no-cache: 16,9246796


with conftests
--------------

python generate_tests_tree.py 4 10 5 1
2000 tests
cache: 2,1642561
no-cache: 2,4153508

python generate_tests_tree.py 4 5 10 1
2000 tests
cache: 2,0458092
no-cache: 2,1860023 

python generate_tests_tree.py 10 10 30 1
30000 tests
cache: 18,6366042
no-cache: 27,1169733
```

The timings were obtained by running `pytest --collect-only` in the generated directory and measuring that using powershell.

So it seems there's some gain by using the cache (specially when there are conftests), but they see minimal in my opinion. That 9-10s difference with **30000** tests doesn't seem relevant, after all with 30000 tests that test suite, if it were real, would take several minutes to run (if not hours). It takes 76s to run the full suite, and those are all tests with just `pass`.

Unless there's something wrong with my testing, I agree that we could just take out that cache.

**Comment by d-b-w** [comment]
Cool! Really nice experiment.

**Comment by hpk42** [comment]
great analysis.  This really calls for removing the cache.  If a caching layer does not provide measurable benefit it should go.  There are two hard CS problems: naming, caching and off-by-one errors.

**Comment by nicoddemus** [comment]
Thanks @hpk42! I will update the PR to remove the cache then. 👍

**Comment by nicoddemus** [comment]
Changed target to `master` because the change now is perfectly safe.

**Pull request merged**